sympy/core/numbers.py

```diff
@@ -222,6 +223,63 @@ def _literal_float(f):
     """Return True if n starts like a floating point number."""
     return bool(_floatpat.match(f))
 
+
+def _most_sig(x):
+    """Return the most significant digit's distance from the
+       decimal point of an _mpf_ value.
+    """
+    # return 0 if mpf precision < 1
+    if x[3] < 1:
+        return 0
+
+    x = (0,) + x[1:]
+    i = mlib.libmpf.to_int(x)
+    if 1 > i:
+        # fall back to log for values < 1
+        prec = max(30, x[3])
+        return mlib.to_int(mlib.mpf_div(
+            mlib.mpf_log(x, prec), mlib.mpf_ln10(prec), prec
+        ))
+    # a bit dirty, but is not slower compared to
+    # log and gives a much cleaner result
+    return len(str(i))
+
+
+def _significants(x):
+    """Returns the most significant and least uncertain digits'
+       distance from the decimal point.
+    """
+    most_sig = _most_sig(x._as_mpf_val(x._prec))
+    return most_sig, most_sig - prec_to_dps(x._prec)
+
+
+def _significants_helper(f, other):
+    """Helper function to figure out significant digits and
+       working precision for addition/subtraction b/w Floats.
```

Comment:
I think it would be good to have some doctest examples for these functions to clearly demonstrate what they do. Also we need to clarify everywhere whether we are talking about decimal or binary digits or precision etc.

This function returns precision in decimal digits and uses `prec_to_dps` everywhere. However in all of the places where this function is used the result is immediately converted back with `dps_to_prec`.

Would it not make more sense for this to return binary precision?

Reply:
Well, I did think about implementing it completely in binary, but that would result in rounding off to powers of 2. 
So most calculation has to be done in decimal before being converted to binary.

There are always two `prec_to_dps` calls to convert both numbers' precision to decimal, to calculate significant digits; and then two `dps_to_prec` calls, one calculate the precision at which to perform the actual operation, and another to increase/reduce precision to the final significant digits. There might be a slightly better approach to this (usually there's always a slightly better way to do everything), but this is best I could come up with.

It might also be a good idea to cache the significant digits of a Float after a calculation since it's already known. This will reduce quite a few required calculations when performing more than one addition/subtraction—especially when multiple operations are carried out in succession—but I'm not really sure how to go about doing that.

I'll work on the documentation tho, it is a bit vague.

Reply:
Are the significant digits not stored as part of the float? What exactly is it that would be cached?

Reply:
They are as precision, but to perform addition or subtraction we need to figure out another thing. 
Take a look at something like this:
```
  1 2 3 . 2 6
+   8 6 . 3
 -------------
  2 0 9 . 6

  1 2 3 . 2 6
+ 8 6 3 .
 -------------
  9 8 6 .
```
Note that the operands in both examples have the same precision, but the results are of different precision.

For this you'll also need to figure out the _position_ of the most and least significant digits, to round off to the required precision. Since the position of the most and least significant digits of the result is also calculated to figure out the precision, they can be cached to avoid recalculation when another operation is performed on the Float.

But like I mentioned previously I'm not sure as to how to cache these "positions" in accordance with how things work with sympy. Normally I would do something unimpressive like making a new attribute and storing it in there, but that doesn't feel right.

Reply:
> Well, I did think about implementing it completely in binary, but that would result in rounding off to powers of 2.
> So most calculation has to be done in decimal before being converted to binary when it's required.

I think that this is backwards. The calculations are done in binary and should be done in binary. Note that binary has advantageous rounding properties over decimal. All the calls to `dps_to_prec` and `prec_to_dps` indicate that the `_most_sig` etc functions are working in the wrong base.

Converting to decimal should only be done when required (when converting to a string).

Reply:
> I think that this is backwards. The calculations are done in binary and should be done in binary. Note that binary has advantageous rounding properties over decimal. All the calls to dps_to_prec and prec_to_dps indicate that the _most_sig etc functions are working in the wrong base.

Well, I don't really see how you could round to decimals without a radix conversion. 

I tried looking in sympy's own round function, but that also has a similar approach:
https://github.com/sympy/sympy/blob/6acf7eadb39677fa728ac6437339f4d90c33d961/sympy/core/expr.py#L3726-L3729

The `_mag` function is pretty much the same as `_most_sig`, but uses ceiling instead of a floor to round to an integer.

Although instead of just reducing precision, round seems to shift all the numbers to the right of the decimal point, converts to an int, rounds as an integer, and then shifts the digits back.

If you could point me to a resource or explain to me how you could implement this in binary (in brief) that would be great.

Reply:
> Well, I don't really see how you could round to decimals without a radix conversion.

We shouldn't round to decimals (except possibly when printing). The rounding should be done to a number of binary digits. I would expect mpmath to have a function for doing this.